velox/expression/tests/ExprTest.cpp

```diff
@@ -3263,3 +3263,52 @@ TEST_F(ExprTest, flatNoNullsFastPathWithCse) {
   assertEqualVectors(expected, result);
   EXPECT_EQ(5, stats.at("plus").numProcessedRows);
 }
+
+TEST_F(ExprTest, caseOverLazyDictionary) {
+  auto input = makeRowVector({
+      makeConstant<int64_t>(10, 5),
+      std::make_shared<LazyVector>(
+          pool(),
+          BIGINT(),
+          5,
+          std::make_unique<SimpleVectorLoader>([=](RowSet /*rows*/) {
+            return wrapInDictionary(
+                makeIndicesInReverse(5),
+                makeFlatVector<int64_t>({8, 9, 10, 11, 12}));
+          })),
+      makeFlatVector<int64_t>({1, 2, 10, 11, 12}),
+  });
+
+  // if (c1 > 10, c0 + c1, c0 - c1) is a null-propagating conditional CSE.
+  auto result = evaluate(
+      "if (c2 > 10::bigint, "
+      "   if (c1 > 10, c0 + c1, c0 - c1) + c2, "
+      "   if (c1 > 10, c0 + c1, c0 - c1) - c2)",
+      input);
+
+  auto expected = makeFlatVector<int64_t>({21, 19, -10, 12, 14});
+  assertEqualVectors(expected, result);
+}
+
+TEST_F(ExprTest, cseOverDictionary) {
```

Comment:
can you also add the following tests:
1. the inputs are constant and peeling generates a const peel.
2. Another case with a mix of const and dictionary inputs.
3. Case where the second time CSE is employed 'rows' is not a subset of 'sharedRows'

Reply:
@bikramSingh91 These are good suggestions. Adding the tests. #3 discovered the bug you reported in #3958. Will fix that too.